app/src/main/java/org/dhis2/Bindings/ValueExtensions.kt

```diff
@@ -3,27 +3,158 @@ package org.dhis2.Bindings
 import org.hisp.dhis.android.core.D2
 import org.hisp.dhis.android.core.common.ValueType
 import org.hisp.dhis.android.core.trackedentity.TrackedEntityAttributeValue
+import org.hisp.dhis.android.core.trackedentity.TrackedEntityAttributeValueObjectRepository
+import org.hisp.dhis.android.core.trackedentity.TrackedEntityDataValue
+import org.hisp.dhis.android.core.trackedentity.TrackedEntityDataValueObjectRepository
 
 fun TrackedEntityAttributeValue.userFriendlyValue(d2: D2): String? {
-    val attribute = d2.trackedEntityModule()
-        .trackedEntityAttributes().uid(trackedEntityAttribute())
+    if (value().isNullOrEmpty())
+        return value()
+
+    val attribute = d2.trackedEntityModule().trackedEntityAttributes()
+        .uid(trackedEntityAttribute())
         .blockingGet()
 
-    if(value().isNullOrEmpty())
+    if (check(d2, attribute.valueType(), attribute.optionSet()?.uid(), value()!!)) {
+        attribute.optionSet()?.let {
+            return checkOptionSetValue(d2, it.uid(), value()!!)
+        } ?: return checkValueTypeValue(d2, attribute.valueType(), value()!!)
+    } else {
+        return null
+    }
+}
+
+fun TrackedEntityDataValue.userFriendlyValue(d2: D2): String? {
+    if (value().isNullOrEmpty())
         return value()
 
-    attribute.optionSet()?.let {
-        val option = d2.optionModule().options()
-            .byOptionSetUid().eq(it.uid())
-            .byCode().eq(value()).one().blockingGet()
-        return if(option!=null) {
-            option.displayName()
-        }else{
-            "Invalid value"
+    val dataElement = d2.dataElementModule().dataElements()
+        .uid(dataElement())
+        .blockingGet()
+
+    if (check(d2, dataElement.valueType(), dataElement.optionSet()?.uid(), value()!!)) {
+        dataElement.optionSet()?.let {
+            return checkOptionSetValue(d2, it.uid(), value()!!)
+        } ?: return checkValueTypeValue(d2, dataElement.valueType(), value()!!)
+    } else {
+        return null
+    }
+}
+
+fun checkOptionSetValue(d2: D2, optionSetUid: String, code: String): String? {
+    return d2.optionModule().options()
+        .byOptionSetUid().eq(optionSetUid)
+        .byCode().eq(code).one().blockingGet()?.displayName()
+}
+
+fun checkValueTypeValue(d2: D2, valueType: ValueType?, value: String): String {
+    return when (valueType) {
+        ValueType.ORGANISATION_UNIT ->
+            d2.organisationUnitModule().organisationUnits()
+                .uid(value)
+                .blockingGet()
+                .displayName()!!
```

Comment:
Can display name be null ? If so it will crash (nobody catch the exception)

Reply:
Nop, displayName is never null, this may be one of the wrong nullable tags in the sdk models